MIG-1030: Document downstream network policy configuration compatibility with DVM (#45431)
openshift/openshift-docs: modules/migration-about-configuring-proxies.adoc

// Module included in the following assemblies:
//
// * migrating_from_ocp_3_to_4/installing-3-4.adoc
// * migrating_from_ocp_3_to_4/installing-restricted-3-4.adoc
// * migration_toolkit_for_containers/installing-mtc.adoc
// * migration_toolkit_for_containers/installing-mtc-restricted.adoc

:_content-type: CONCEPT
[id="migration-about-configuring-proxies-{context}"]
= Proxy configuration

For {product-title} 4.1 and earlier versions, you must configure proxies in the `MigrationController` custom resource (CR) manifest after you install the {mtc-full} Operator because these versions do not support a cluster-wide `proxy` object.

For {product-title} 4.2 to {product-version}, the {mtc-full} ({mtc-short}) inherits the cluster-wide proxy settings. You can change the proxy parameters if you want to override the cluster-wide proxy settings.

> == Direct volume migration

Author: Fixed

> Direct Volume Migration (DWM) was introduced in MTC 1.4.2. DVM supports only one proxy. The source cluster cannot access the route of the target cluster if the target cluster is also behind a proxy.

Reviewer: s/DWM/DVM

> If you want to perform a DVM from a source cluster behind a proxy, you must configure a TCP proxy that works at the transport layer and forwards the SSL connections transparently without decrypting and re-encrypting them with their own SSL certificates. A Stunel proxy is an example of such a proxy.

Reviewer 2: Stunel -> Stunnel

> === TCP proxy setup for DVM

Author: Fixed.

> You can set up direct connection between the source and the target cluster through a TCP proxy and configure the `stunnel_tcp_proxy` variable in the `MigrationController` CR use the proxy:

Reviewer 3: Is there a word missing e.g. **to** between CR and use

Author: Fixed.

[source, yaml]
----
apiVersion: migration.openshift.io/v1alpha1
kind: MigrationController
metadata:
  name: migration-controller
  namespace: openshift-migration
spec:
  [...]
  stunnel_tcp_proxy: http://username:password@ip:port
----

> Direct volume migration (DVM) only supports Basic Authentication for the proxy. Moreover, DVM only works behind proxies which can tunnel a TCP connection transparently. HTTP/HTTPS proxies in man-in-the-middle mode will not work. The existing cluster-wide proxies may or may not support this behavior. As a result, the proxy settings for DVM are intentionally kept different from usual proxy configuration in {mtc-short}.

Reviewer 3: Basic Authentication >> basic authentication
Add **the** to the line  .....kept different from **the** usual proxy ....

> === Why use a TCP proxy instead of an HTTP/HTTPS proxy?

Author: Fixed.

> DVM is enabled by running Rsync between the source and the target cluster over an OpenShift Route.  Traffic is encrypted using Stunnel, which in itself is a TCP proxy. The Stunnel running on the source cluster initiates a TLS connection with the target Stunnel and transfers data over an encrypted channel.

Reviewer 3: OpenShift Route >> OpenShift route

Author: Fixed

> DVM is enabled by running Rsync between the source and the target cluster over an OpenShift route.  Traffic is encrypted using Stunnel, which in itself is a TCP proxy. The Stunnel running on the source cluster initiates a TLS connection with the target Stunnel and transfers data over an encrypted channel.

Author: Fixed.

> Cluster-wide HTTP/HTTPS proxies in OpenShift are usually configured in man-in-the-middle mode where they negotiate their own TLS session with the outside servers. However, this does not work with Stunnel. Stunnel requires that its TLS session be untouched by the proxy, essentially making the proxy a transparent tunnel which simply forwards the TCP connection as-is. Therefore, a TCP proxy must be used.

Author: Fixed.

=== Known Issue

.Migration fails with error `Upgrade request required`

> The migration Controller uses the SPDY protocol to execute commands within remote Pods. If the remote cluster is behind a proxy or a firewall that does not support the SPDY protocol, the migration controller  fails to execute remote commands. The migration fails with the error message `Upgrade request required`.

Reviewer 3: The migration Controller uses the SPDY protocol to execute commands within remote Pods.

- Please expand **SPDY** in the first usage in the module.
- Pods >> pods (https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#api-object-formatting)

Reviewer 3: the migration controller  fails to execute remote commands >>please check spacing between **controller** and **fails.**

Author: SPDY is the protocol's name. It is not an acronym.  "The name "SPDY" is a [trademark](https://en.wikipedia.org/wiki/Trademark)[[6]](https://en.wikipedia.org/wiki/SPDY#cite_note-6) of Google and is not an [acronym](https://en.wikipedia.org/wiki/Acronym)" from https://en.wikipedia.org/wiki/SPDY.

Author: Other changes made.

Workaround: Use a proxy that supports the SPDY protocol.

> In addition to supporting the SPDY protocol, the proxy or firewall also must pass the `Upgrade` HTTP header to the API server. The client uses this header to open a websocket connection with the API server. If the `Upgrade` header is blocked by the proxy or firewall, the migration fail with the error message `Upgrade request required`.

Author: Fixed.

Workaround: Ensure that the proxy forwards the `Upgrade` header.

> == Tuning Network Policies for migrations

Reviewer 3: Use sentence case as per (https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#assemblymodule-titles-and-section-headings)  >>Tuning network policies for migrations

Author: Fixed.

> OpenShift supports restricting traffic to or from Pods using _NetworkPolicy_ or and _EgressFirewalls_ based on the network plugin used by the cluster. If any of the source namespaces involved in a migration use such mechanisms to restrict network traffic to Pods, the restrictions might inadvertently stop traffic to Rsync Pods during migration.

Reviewer 3: Please replace instances of **Pods** with **pods**

> OpenShift supports restricting traffic to or from Pods using _NetworkPolicy_ or and _EgressFirewalls_ based on the network plugin used by the cluster. If any of the source namespaces involved in a migration use such mechanisms to restrict network traffic to Pods, the restrictions might inadvertently stop traffic to Rsync Pods during migration.

Reviewer 3: using _NetworkPolicy_ or **and** _EgressFirewalls_ based >> is the and required?

 Rsync Pods running on both the source and the target clusters must connect to each other over an OpenShift Route. Existing _NetworkPolicy_ or _EgressNetworkPolicy_ objects can be configured to automatically exempt Rsync Pods from these traffic restrictions.

=== NetworkPolicy configuration

> ==== Egress traffic from Rsync Pods

Reviewer 3: Please use sentence case (https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#assemblymodule-titles-and-section-headings)

Author: Problem solved when I switched "Pods" to "pods." It wasn't a sentence case issue.

> You can use the unique labels of Rsync pods to allow egress traffic from them if the `NetworkPolicy` configuration in source or destination namespaces blocks this type of traffic. The following policy allows *all* egress traffic from Rsync pods in the namespace:

Author: Fixed.

[source, yaml]
----
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-all-egress-from-rsync-pods
spec:
  podSelector:
    matchLabels:
      owner: directvolumemigration
      app: directvolumemigration-rsync-transfer
  egress:
  - {}
  policyTypes:
  - Egress
----

> ==== Ingress traffic to Rsync Pods

Reviewer 3: sentence case

> ==== Ingress traffic to Rsync pods

Author: Fixed.

[source, yaml]
----
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-all-egress-from-rsync-pods
spec:
  podSelector:
    matchLabels:
      owner: directvolumemigration
      app: directvolumemigration-rsync-transfer
  ingress:
  - {}
  policyTypes:
  - Ingress
----

> === EgressNetworkPolicy configuration

Author: Fixed

> `EgressNetworkPolicy` or _Egress Firewalls_ are OpenShift constructs designed to block egress traffic leaving the cluster.

Author: Fixed.

> Unlike `NetworkPolicy`, the Egress Firewall works at a project level in that it applies to all pods in the namespace. Therefore, the unique labels of Rsync pods cannot be used to exempt only Rsync pods from the restrictions. However, you can add the CIDR ranges of the source or target cluster to the _Allow_ rule of the policy so that a direct connection can be setup between two clusters.

Reviewer 4: Small tweak for active voice and clarity:
```suggestion
Unlike the `NetworkPolicy` object, the Egress Firewall works at a project level because it applies to all pods in the namespace. Therefore, the unique labels of Rsync pods do not exempt only Rsync pods from the restrictions. However, you can add the CIDR ranges of the source or target cluster to the _Allow_ rule of the policy so that a direct connection can be setup between two clusters.
```

Author: Fixed.

> Based on which cluster the Egress Firewall is present in, you can add the CIDR range of the other cluster to allow _egress_ traffic between the two:

Reviewer 4: Suggestion to remove italics:
```suggestion
Based on which cluster the Egress Firewall is present in, you can add the CIDR range of the other cluster to allow egress traffic between the two:
```
There's a previous instance where "egress" is not italicized, and I'm not sure if it requires emphasis here...if it does though, please feel free to disregard 🙂

Author: Fixed.
